lib/importer/backend.js

```diff
@@ -0,0 +1,300 @@
+    // the preview, so that users can see their spreadsheet in a more familiar
+    // form, and because styling information might be a significant part of the
+    // data.
+
+    // Slice out only desired columns from these rows, and return the results
+    return [startRows.map((row) => row.slice(range.start.column, range.end.column+1)),
+            middleRows.map((row) => row.slice(range.start.column, range.end.column+1)),
+            endRows.map((row) => row.slice(range.start.column, range.end.column+1))];
+};
+
+// Return the unique values in each column in the range. Return no more than
+// maxValues values for any given column. Return format is an array, one entry
+// per column, whose entries have a .values property that's an array of values
+// and a .hasMore property that's a boolean set if the values array was
+// truncated to maxValues.
+exports.SessionGetInputValues = (sid, range, maxValues) => {
+    assert(sessionStore[sid]);
+    assert(sessionStore[sid].sheets[range.sheet]);
+    assert(range.end.row >= range.start.row);
+    assert(range.end.column >= range.start.column);
+
+    let data = sessionStore[sid].sheets[range.sheet];
+    let result = new Array(range.end.column - range.start.column + 1);
+    let resultIdx = 0;
+
+    for(let col = range.start.column; col <= range.end.column; col++) {
+        let values = new Set();
+        let hasMore = false;
+
+        for(let row = range.start.row; row <= range.end.row; row++) {
+            let value = data[row][col];
+
+            if (values.size < maxValues) {
+                // There's room for more, keep shovelling them in
+                values.add(value);
+            } else {
+                // It's full - but is this a NEW value?
+                if (!values.has(value)) {
+                    // Yes, it's a new value, so set the "hasMore" flag and stop iterating down the column
+                    hasMore = true;
+                    break;
+                }
+            }
+        }
+
+        // Convert the result to a sorted array, for neatness and consistency
+        result[resultIdx] = { values: Array.from(values.values()).sort(),
+                              hasMore: hasMore };
+        resultIdx++;
+    }
+
+    return result;
+}
+
+exports.SessionDelete = (sid) => {
+    assert(sessionStore[sid]);
+    sessionStore[sid].jobs.forEach((jid) => {
+        delete jobStore[jid];
+    });
+    delete sessionStore[sid];
+}
+
+exports.SessionPerformMappingJob = (sid, range, mapping) => {
+    assert(sessionStore[sid]);
+    assert(sessionStore[sid].sheets[range.sheet]);
+    assert(range.end.row >= range.start.row);
+    assert(range.end.column >= range.start.column);
+    // FIXME: More asserts that the mapping makes sense
+
+    let job = {
+        records: [], // Array of output records
+        errors: new Map(), // Maps input row number to list of errors
+        warnings: new Map() // Maps input row number to list of warnings
+    };
```

Comment:
I think I've done this a lot as well, but I wonder whether we would be better to move this somewhere so that we have something that can act as 'documentation' for what a job looks like.  I don't know whether a `makeJob()` or a class/constructor would make most sense, but it might make the contents more obvious?

Reply:
Ok!